### PR TITLE
[BEAM-248] Add display data to additional PTransforms

### PR DESCRIPTION
**Pull request by swegner**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by swegner** [comment]
R: @bjchambers

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/gcsfs/GcsPath.java [hunk 1]
```diff
@@ -601,6 +600,27 @@ public String toResourceName() {
     return sb.toString();
   }
 
+  private static final Pattern HAS_GLOB = Pattern.compile(".*(\\[|\\*|\\?).*");
+  public String getBrowseUrl() {
+    StringBuilder builder = new StringBuilder("https://console.cloud.google.com/storage/browser/");
```
Comment (swegner) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/gcsfs/GcsPath.java]: It appears these URLs work for browsing to "buckets" (i.e. directories), but not "objects" (i.e. files). There's another URL schema, "https://storage.cloud.google.com/*" which works for exactly the opposite-- buckets but not objects.  Without doing a query, we won't always be able to tell what's a bucket and what's an object.

Comment (swegner) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/gcsfs/GcsPath.java]: See: https://cloud.google.com/storage/docs/cloud-console#_accessing

Comment (swegner) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/gcsfs/GcsPath.java]: I've split off link URLs to a separate PR: https://github.com/apache/incubator-beam/pull/300

**Comment by swegner** [comment]
I've addressed all feedback so far. Please take another look. @bjchambers

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/transforms/DataflowDisplayDataEvaluator.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  /** Do not instantiate. */
+  private DataflowDisplayDataEvaluator() {}
+
+  /**
+   * Create a {@link DisplayDataEvaluator} instance to evaluate pipeline display data against
+   * the {@link DataflowPipelineRunner}.
+   *
+   * @param tempLocation Temp location for {@link DataflowPipelineOptions#setTempLocation(String)}
+   */
+  static DisplayDataEvaluator create(File tempLocation) {
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/transforms/DataflowDisplayDataEvaluator.java]: If this is only used for testing, I'd propose taking a `TemporaryFile` instance directly. That will force the tests to be properly written to use `@Rule TemporaryFile tempFile` and then call `DisplayDataEvaluator.create(tempFile)`. Other option which would be even more extreme is to make this a `@Rule`.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/transforms/DataflowTextIOTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+/**
+ * Unit tests for Dataflow usage of {@link TextIO} transforms.
+ */
+public class DataflowTextIOTest {
+  @Rule
+  public TemporaryFolder tempFolder = new TemporaryFolder();
+
+  @Test
+  public void testPrimitiveDisplayData() {
+    DisplayDataEvaluator evaluator = DataflowDisplayDataEvaluator.create(tempFolder.getRoot());
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/transforms/DataflowTextIOTest.java]: Proposal:
```
@Rule
public DisplayDataEvaluator evaluator = new DataflowDisplayDataEvaluator();

@Test
public void testPrimitiveDisplayData() {
  ...
  evaluator.displayDataForPrimitiveTransform(write);
}
```

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/display/DisplayDataEvaluator.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+      .apply(input)
+      .apply(root);
+
+    PrimitiveDisplayDataPTransformVisitor visitor = new PrimitiveDisplayDataPTransformVisitor(root);
+    pipeline.traverseTopologically(visitor);
+    return visitor.getDisplayDatas();
+  }
+
+  /**
+   * Visits {@link PTransform PTransforms} in a pipeline, and collected {@link DisplayData} for
```
Comment (bjchambers) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/display/DisplayDataEvaluator.java]: "collected" -> "collects"

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/transforms/DataflowDisplayDataEvaluator.java (hunk 2 above)
Comment (swegner) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/transforms/DataflowDisplayDataEvaluator.java]: The temp value parameter is gone in the latest version. Some DataflowPipelineOptions require a GCS path, so we're using a stub GCS instead.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/transforms/DataflowTextIOTest.java (hunk 3 above)
Comment (swegner) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/transforms/DataflowTextIOTest.java]: This would mean having the evaluator make the assertions as well, right? Then the syntax would be something like:

    @Test
    public void testPrimitiveDisplayData() {
      ...
      evaluator.displayDataForPrimitiveTransform(write, 
          hasDisplayItem(hasValue(startsWith("foobar"))));
    }

That might be ok. It makes the usage more specific, but we only have one usage scenario now anyway. I don't want to add too much layering to the usage so it's easy to understand. I'll think about it more..

**Comment by swegner** [comment]
I've addressed all feedback so far. Please take another look. @bjchambers

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/io/DataflowBigQueryIOTest.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+public class DataflowBigQueryIOTest {
+  @Test
+  public void testBatchSinkPrimitiveDisplayData() {
+    DataflowPipelineOptions options = DataflowDisplayDataEvaluator.getDefaultOptions();
+    options.setStreaming(false);
+    testSinkPrimitiveDisplayData(options);
+  }
+
+  @Test
+  public void testStreamingSinkPrimitiveDisplayData() {
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/io/DataflowBigQueryIOTest.java]: I think for Junit4 it is preferred not to prefix with "test". So just"
```
@Test
public void streamingSinkPrimitiveDisplayData
```

etc.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/io/DataflowBigQueryIOTest.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  private void testSinkPrimitiveDisplayData(DataflowPipelineOptions options) {
+    DisplayDataEvaluator evaluator = DataflowDisplayDataEvaluator.create(options);
+
+    BigQueryIO.Write.Bound write = BigQueryIO.Write
+        .to("project:dataset.table")
+        .withSchema(new TableSchema().set("col1", "type1").set("col2", "type2"))
+        .withoutValidation();
+
+    Set<DisplayData> displayDatas = evaluator.displayDataForPrimitiveTransforms(write);
+    assertThat(displayDatas, hasItem(hasDisplayItem(hasKey("tableSpec"))));
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/io/DataflowBigQueryIOTest.java]: Use of `hasItem` (from hamcrest) here should be read as `hasStep`, right? Seems fine, just confirming.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/io/DataflowTextIOTest.java [hunk 7]
```diff
@@ -116,4 +126,14 @@ public void testGoodWildcards() throws Exception {
   private void applyRead(Pipeline pipeline, String path) {
     pipeline.apply("Read(" + path + ")", TextIO.Read.from(path));
   }
+
+  @Test
+  public void testPrimitiveDisplayData() {
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/io/DataflowTextIOTest.java]: Ditto on test naming.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/io/DataflowBigQueryIOTest.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+  private void testSinkPrimitiveDisplayData(DataflowPipelineOptions options) {
+    DisplayDataEvaluator evaluator = DataflowDisplayDataEvaluator.create(options);
+
+    BigQueryIO.Write.Bound write = BigQueryIO.Write
+        .to("project:dataset.table")
+        .withSchema(new TableSchema().set("col1", "type1").set("col2", "type2"))
+        .withoutValidation();
+
+    Set<DisplayData> displayDatas = evaluator.displayDataForPrimitiveTransforms(write);
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/io/DataflowBigQueryIOTest.java]: `displayData` is already plural, so not strictly necessary to call this `displayDatas`.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/transforms/DataflowCombineTest.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    DisplayDataEvaluator evaluator = DataflowDisplayDataEvaluator.create();
+
+    CombineTest.UniqueInts combineFn = new CombineTest.UniqueInts();
+    PTransform<PCollection<KV<Integer, Integer>>, ? extends POutput> combine =
+        Combine.perKey(combineFn);
+
+    Set<DisplayData> displayDatas = evaluator.displayDataForPrimitiveTransforms(combine,
+        KvCoder.of(VarIntCoder.of(), VarIntCoder.of()));
+
+    assertThat(displayDatas, hasItem(hasDisplayItem("combineFn", combineFn.getClass())));
```
Comment (bjchambers) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/transforms/DataflowCombineTest.java]: Maybe a comment on each of these tests emphasizing what it is testing.
`// Ensure that a Combine.perKey includes the combineFn being used.`

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/ProxyInvocationHandler.java [hunk 10]
```diff
@@ -281,6 +281,10 @@ private void populateDisplayData(DisplayData.Builder builder) {
       HashSet<PipelineOptionSpec> specs = new HashSet<>(optionsMap.get(option.getKey()));
 
       for (PipelineOptionSpec optionSpec : specs) {
+        if (!optionSpec.shouldSerialize()) {
```
Comment (bjchambers) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/ProxyInvocationHandler.java]: Can you add a comment here and below on what this check on `shouldSerialize` is doing? It is unclear why we would want to skip options.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/MapElementsTest.java [hunk 11]
```diff
@@ -103,7 +106,7 @@ public String apply(String input) {
     assertThat(pipeline.getCoderRegistry().getDefaultCoder(output.getTypeDescriptor()),
         equalTo(pipeline.getCoderRegistry().getDefaultCoder(new TypeDescriptor<String>() {})));
 
-    // Make sure the pipelien runs too
+    // Make sure the  pipelien runs too
```
Comment (bjchambers) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/MapElementsTest.java]: `Make sure the  pipelien` -> Make sure the pipeline`

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/MapElementsTest.java [hunk 12]
```diff
@@ -118,12 +121,41 @@ public void testVoidValues() throws Exception {
...
+    assertThat(DisplayData.from(simpleMap), hasDisplayItem("mapFn", simpleFn.getClass()));
+  }
+
   static class VoidValues<K, V>
       extends PTransform<PCollection<KV<K, V>>, PCollection<KV<K, Void>>> {
 
     @Override
     public PCollection<KV<K, Void>> apply(PCollection<KV<K, V>> input) {
-      return input.apply(MapElements.<KV<K, V>, KV<K, Void>>via(
+      return input.apply(MapElements.via(
```
Comment (bjchambers) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/MapElementsTest.java]: Make sure you've tested this with a JDK 7 -- sometimes these are necessary for older version support.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/display/DisplayDataEvaluator.java [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+   * Traverse the specified {@link PTransform}, collecting {@link DisplayData} registered on the
+   * inner primitive {@link PTransform PTransforms}.
+   *
+   * @param root The root {@link PTransform} to traverse
+   * @param inputCoder The coder to set for the {@link PTransform} input, or null to infer the
+   *                   default coder.
+   *
+   * @return the set of {@link DisplayData} for primitive {@link PTransform PTransforms}.
+   */
+  public <InputT> Set<DisplayData> displayDataForPrimitiveTransforms(
```
Comment (bjchambers) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/display/DisplayDataEvaluator.java]: @kennknowles Should this be for "primitive transforms" or "leaf transforms"?

**Comment by bjchambers** [comment]
Some comments, then LGTM.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/display/DisplayDataEvaluator.java (hunk 13 above)
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/display/DisplayDataEvaluator.java]: Interesting question/bikeshed :-). I just altered `PipelineVisitor` to call them primitive, but leaf is probably better.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/io/DataflowBigQueryIOTest.java (hunk 5 above)
Comment (swegner) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/io/DataflowBigQueryIOTest.java]: The convention is split in the BEAM codebase, but most test cases use the "test" prefix: 1916 prefixed vs 93 not according to [this script](https://gist.github.com/swegner/d6425e703d10d82600d0e98319257c82).

I'm going to follow the BEAM convention. It makes sense to migrate if we're not following JUnit4 convention, but we should do it across the codebase. Feel free to file a bug assigned to me if you feel it's worth the effort.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/io/DataflowBigQueryIOTest.java (hunk 6 above)
Comment (swegner) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/io/DataflowBigQueryIOTest.java]: Correct. In most other cases I've been asserting on `DisplayData` from a single transform, but `DisplayDataEvaluator` returns a set of `DisplayData` from all of its primitives, so this asserts over any of them.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/transforms/DataflowCombineTest.java (hunk 9 above)
Comment (swegner) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/transforms/DataflowCombineTest.java]: Good call. Adding it to the assert method.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/MapElementsTest.java (hunk 12 above)
Comment (swegner) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/MapElementsTest.java]: Got it. Our Travis gauntlet has a few different JDKs and didn't flag anything.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/display/DisplayDataEvaluator.java (hunk 13 above)
Comment (swegner) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/display/DisplayDataEvaluator.java]: We use "primitive transform" as a concept in a few places, but I didn't find instances of "leaf".

Does "leaf" accurately describe the concept? My understanding is that a primitive transform is an injection point where the runner may replace the node with a different implementation. Can the injected implementation include other transforms?

**Comment by swegner** [comment]
I've addressed all feedback so far. Please take another look. @bjchambers

**Comment by bjchambers** [comment]
LGTM

**Comment by swegner** [comment]
Backport: https://github.com/GoogleCloudPlatform/DataflowJavaSDK/pull/230
